xarray/core/datatree.py

```diff
@@ -1042,10 +1064,12 @@ def drop_nodes(
             if extra:
                 raise KeyError(f"Cannot drop all nodes - nodes {extra} not present")
 
+        result = self.copy()
         children_to_keep = {
-            name: child for name, child in self.children.items() if name not in names
+            name: child for name, child in result.children.items() if name not in names
         }
-        return self._replace(children=children_to_keep)
+        result._replace_node(children=children_to_keep)
+        return result
```

Comment:
oh this is another example of "functions should either have a return value or operate in-place, not both" isn't it?

Reply:
Yes, exactly!